[backport] chore: Add QueuedMessageResponse in epoching module(babylonlabs-io#1347)

jeoghyukLee123 opened the pull request:
# Description

Closes: #1312 

Adds a new enriched_msg field to QueuedMessageResponse, allowing clients to distinguish between messages like MsgDelegate and MsgUndelegate even if their raw msg field looks identical.

## Notes

Adds support for a structured EnrichedMsg field in the QueuedMessageResponse so clients can distinguish between otherwise identical raw messages.

- Introduce EnrichedMsg in the protobuf definition and disable automatic getters.
- Populate EnrichedMsg in Go’s ToResponse helper and update generated gRPC code.
- Extend tests to cover the new field and adjust CLI code for removed getters.

wnjoon commented:
Commit Summary

[fd5e873](https://github.com/b-harvest/babylon/pull/8/commits/fd5e873ebe84d30737e00c2afddad853af32a1da)
- Recreate the WrappedFilePV structure with unnecessary fields removed.
- Change BlsSigner interface implementation back to WrappedFilePV.

[946eab5](https://github.com/b-harvest/babylon/pull/8/commits/946eab5aa98da04402b3beced99d93c8fcd232d3)
- Add and test `InitCmd` function that wraps cosmos' InitCmd.
- Use the function used in create-bls-key to generate a bls key.

[57ddd51](https://github.com/b-harvest/babylon/pull/8/commits/57ddd51657d902a3b096df70482399a639d8e067)
- Remove DelegatorAddress validation when calling ExtendVote function

[280de7e](https://github.com/b-harvest/babylon/pull/8/commits/280de7e65f206d04a763d5de0a956984235a4e7e)
- Check if key file exists in `InitPrivSigner` when starting node and output action message

dongsam commented on x/checkpointing/vote_ext.go:
```diff
@@ -81,7 +80,7 @@ func (h *VoteExtensionHandler) ExtendVote() sdk.ExtendVoteHandler {
 		ve := &ckpttypes.VoteExtension{
 			Signer:           signer.String(),
-			ValidatorAddress: k.GetValidatorAddress().String(),
+			ValidatorAddress: signer.String(),
```
Previously, `GetValidatorAddress()` returned a value of type `sdk.ValAddress` for the validator pubkey. However, it now returns `sdk.ValAddress` for the delegator address (valoper). The original approach should have used the [valconspub (ConsAddress)](https://github.com/cosmos/cosmos-sdk/blob/ec58adf8e2fa4149c093d871d2fd398ca508525a/types/address.go#L491) type instead of [sdk.ValAddress](https://github.com/cosmos/cosmos-sdk/blob/ec58adf8e2fa4149c093d871d2fd398ca508525a/types/address.go#L337) for the validator pubkey, but it appears this was implemented incorrectly. Consider leaving this as a separate issue or discussing it with the team. Until it is finalized, ensure consistency in address types with the original approach.

dongsam commented on x/epoching/keeper/staking_functions.go:
```diff
@@ -115,3 +115,7 @@ func (k Keeper) CheckMsgCreateValidator(ctx context.Context, msg *stakingtypes.M
 	return k.stk.GetPubKeyByConsAddr(ctx, consAddr)
 }
+
+func (k Keeper) GetValidatorByConsAddr(ctx context.Context, consAddr sdk.ConsAddress) (stakingtypes.Validator, error) {
```
### Consider Removing `GetValidatorPubkey` and `FilePVKey`

- If the `BLS ValidatorPubkey` exists, we can retrieve the associated `ValidatorPubKey` through its pairing (via [`blsKeysToAddr`](https://github.com/babylonlabs-io/babylon/blob/280de7e65f206d04a763d5de0a956984235a4e7e/x/checkpointing/keeper/registration_state.go#L19-L20)).
- This would eliminate the need for `GetValidatorPubkey`.
- Consequently, the [FilePVKey from Comet](https://github.com/babylonlabs-io/babylon/blob/280de7e65f206d04a763d5de0a956984235a4e7e/privval/file.go#L14) in `WrappedFilePV` would also become unnecessary and could be removed.
- Removing `FilePVKey` would reduce overall complexity.

### Observations

- Reading `FilePVKey` at the app level seems unnatural.
- In cases where CometBFT uses a `PrivValidator` that does not rely on `FilePV` (e.g., [Horcrux](https://github.com/strangelove-ventures/horcrux) or other [SignerClient](https://github.com/cometbft/cometbft/blob/3f73754df07fe5155812f7cce001158634cf51dd/privval/signer_client.go#L71)), `FilePV` will not exist.
- Without `FilePV`, it may not be possible to perform vote extensions.

### Benefits of Removing `FilePV`

- Simplifies the architecture by removing dependency on `FilePVKey`.
- Aligns better with modern `PrivValidator` implementations that utilize external signing solutions like `Horcrux`.

dongsam commented:
@wnjoon The base branch of this PR currently overlaps with the branch for [#396](https://github.com/babylonlabs-io/babylon/pull/396). To prevent branches from different PRs from mixing, please fetch https://github.com/babylonlabs-io/babylon/tree/feat/bls-keystore-improvement into the b-harvest repository and update the base branch of this PR

dongsam commented:
@wnjoon Could you add plan on description for migration cmd? 
```
1. the separation feature would be inserted in a new rc release
2. validator stops the current node
3. validator download the new release
4. validator run the migration cmd
5. restart the node and voting should be resumed
```

wnjoon commented:
Commit Summary

[cd6ffcf](https://github.com/b-harvest/babylon/pull/8/commits/cd6ffcf7013774958e3c30aec9077e26987d78c7)

- Add migration cmd

dongsam commented on cmd/babylond/cmd/migrate.go:
```diff
@@ -0,0 +1,93 @@
+
+func MigrateCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "migrate",
```
The [migrate](https://github.com/cosmos/cosmos-sdk/blob/6dc6e8b2a882cd6e2c0e7440c816f2390ee87ee3/x/genutil/client/cli/migrate.go#L34-L44) command is already being used in the existing Babylon version, so it’s necessary to change the cmd name to avoid conflicts. e.g., `migrate-bls-key`

wnjoon replied:
How about `migrate-pv-key`? Since comet and bls keys are separated into different files while migrating, I think `pv` can explain the process exactly.

dongsam replied:
Since the split/migrate process involves the key pair generated by `create-bls-key`, BLS appears to be the main focus rather than PV. Additionally, to maintain consistency in command naming, considering it follows the `create-bls-key` command, `migrate-bls-key` seems more intuitive and understandable IMO

wnjoon replied:
Fixed commit : [4e41051](https://github.com/b-harvest/babylon/pull/8/commits/4e410512c62cbaeb9f877078ea853b6a60c4cd63)

dongsam commented on cmd/babylond/cmd/migrate.go:
```diff
@@ -0,0 +1,93 @@
+		which contained both the bls and comet keys used in previous versions, into separate files.
+
+BLS keys are stored along with other validator keys in priv_validator_key.json in previous version,
+which should exist before running the command (via babylond init or babylond testnet).
```
It would be safer to include a note advising users to back up the priv_validator_key.json file in a secure location before migration to prevent potential loss during the migration process.

wnjoon replied:
Fixed commit : [4e41051](https://github.com/b-harvest/babylon/pull/8/commits/4e410512c62cbaeb9f877078ea853b6a60c4cd63)

dongsam commented on cmd/babylond/cmd/migrate_test.go:
```diff
@@ -0,0 +1,120 @@
+		newPvKeyFile := filepath.Join(configDir, "priv_validator_key.json")
+		newBlsKeyFile := filepath.Join(configDir, "bls_key.json")
+		require.FileExists(t, newPvKeyFile)
+		require.FileExists(t, newBlsKeyFile)
```
It would be helpful to include a process to verify that no key changes or corruption occurred after migration by decrypting the ERC2335-encrypted file and ensuring it matches the pre-migration version of the private key.

wnjoon replied:
Fixed commit : [4e41051](https://github.com/b-harvest/babylon/pull/8/commits/4e410512c62cbaeb9f877078ea853b6a60c4cd63)

dongsam replied:
Please add assertion with `verifyAfterMigration` after migration on `TestMigrate`

wnjoon replied:
Added [d055b84](https://github.com/b-harvest/babylon/pull/8/commits/d055b84ac2f85e42dd8ce89b2b037ea933cc7e8b)

wnjoon commented:
Commit Summary

[4e41051](https://github.com/b-harvest/babylon/pull/8/commits/4e410512c62cbaeb9f877078ea853b6a60c4cd63)

- Change cmd to `migrate0bls-key`
- Add verification to check previous keys are equal to keys in new files
- Add a NOTE to back up previously used key files

dongsam commented on cmd/babylond/cmd/migrate.go:
```diff
@@ -0,0 +1,115 @@
+		prevBlsPrivKey,
+		blsPv.Key.PrivKey,
+	); err != nil {
+		return fmt.Errorf("failed to verify after migration: %w", err)
```
If it fails, `cmtPv.Key.Save()` should not be executed. Adjusting the order to call `cmtPv.Key.Save()` after `verifyAfterMigration` would be safer.

wnjoon replied:
Fixed [d055b84](https://github.com/b-harvest/babylon/pull/8/commits/d055b84ac2f85e42dd8ce89b2b037ea933cc7e8b)

wnjoon commented:
Commit Summary

[d055b84](https://github.com/b-harvest/babylon/pull/8/commits/d055b84ac2f85e42dd8ce89b2b037ea933cc7e8b)

- Modify migration logic (save files after verification)
- Add password prompt (missed)
- Add test case to verify new file after migration

jeoghyukLee123 committed:
[backport] chore: enhance the QueuedMsgs API(babylonlabs-io#1347)
                Closes babylonlabs-io#1312
                ---------
                Co-authored-by: JeonghyukLee <jeonghyuk@bharvest.io>
samricotta <samricotta@gmail.com>

canu0205 committed:
[backport] chore: bump Cosmos SDK to v0.53 (#1223) (#1736)

Minimum requirements to bump Cosmos SDK to v0.53.2. All other updates
are optional.

---------

Co-authored-by: Vladislav Varadinov <vladislav.varadinov@gmail.com>

jeoghyukLee123 committed:
Merge branch 'main' into jhBharvest/backport-1347
